src/utils/pluginFilter.ts

```diff
@@ -0,0 +1,130 @@
+import picomatch from 'picomatch';
+import type { StringFilter, StringOrRegExp } from '../rollup/types';
+import { ensureArray } from './ensureArray';
+import { normalize, relative } from './path';
+
+const FALLBACK_TRUE = 1;
+const FALLBACK_FALSE = 0;
+type FallbackValues = typeof FALLBACK_TRUE | typeof FALLBACK_FALSE;
+
+type PluginFilterWithFallback = (input: string) => boolean | FallbackValues;
+
+export type PluginFilter = (input: string) => boolean;
+export type TransformHookFilter = (id: string, code: string) => boolean;
+
+interface NormalizedStringFilter {
+	include?: StringOrRegExp[];
+	exclude?: StringOrRegExp[];
+}
+
+function patternToIdFilter(pattern: StringOrRegExp): PluginFilter {
+	if (pattern instanceof RegExp) {
+		return (id: string) => {
+			const normalizedId = normalize(id);
+			const result = pattern.test(normalizedId);
+			pattern.lastIndex = 0;
+			return result;
+		};
+	}
+	const cwd = process.cwd();
+	const matcher = picomatch(pattern, { dot: true });
+	return (id: string) => {
+		const normalizedId = normalize(relative(cwd, id));
```

Comment:
This `relative(cwd, id)` normalisation was found to be problematic in rolldown (https://github.com/rolldown/rolldown/pull/3972), so I'm going to add the fix made in rolldown.